src/Compilers/Core/Portable/MetadataReference/CachedAssemblySymbolList.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// This is a collection of weakly held <see cref="IAssemblySymbol"/> instances. 
+    ///
+    /// The methods on this type can be safely called from multiple threads without risk of 
+    /// corrupting the internal data structures.
+    /// </summary>
+    internal sealed class CachedAssemblySymbolList
```

Comment:
>CachedAssemblySymbolList [](start = 26, length = 24)

Would it make sense to inherit from ```WeakList<IAssemblySymbol>``` instead to save on an instance allocation?

Reply:
@AlekseyTs no. That would mean the type has a mix of safe and unsafe methods with respect to calling from multiple threads.